scalding-quotation/src/main/scala/com/twitter/scalding/quotation/Projection.scala

```diff
@@ -0,0 +1,143 @@
+package com.twitter.scalding.quotation
+
+sealed trait Projection {
+  def andThen(prop: String, tpe: String): Projection
+}
+
+/**
+ * A reference of a type. If not nested within a `Property`, it means that all fields are used.
+ */
+final case class TypeReference(tpe: String) extends Projection {
+  def andThen(prop: String, tpe: String) =
+    Property(this, prop, tpe)
+
+  override def toString = tpe.split('.').last
+}
+
+/**
+ * A projection property (e.g. `Person.name`)
+ */
+final case class Property(path: Projection, name: String, tpe: String) extends Projection {
+  def andThen(prop: String, tpe: String) =
+    Property(this, prop, tpe)
+
+  override def toString = s"$path.$name"
+}
+
+/**
+ * Utility class to deal with a collection of projections.
+ */
+final class Projections private (val set: Set[Projection]) extends Serializable {
+
+  /**
+   * Returns the projections that are based on `tpe` and limits projections
+   * to only properties that extend from `superClass`.
+   */
+  def of(tpe: String, superClass: Class[_]): Projections = {
+
+    def byType(p: Projection) = {
+      def loop(p: Projection): Boolean =
+        p match {
+          case TypeReference(`tpe`) => true
+          case TypeReference(_) => false
+          case Property(p, _, _) => loop(p)
+        }
+      loop(p)
+    }
+
+    def bySuperClass(p: Projection) = {
+
+      def isSubclass(c: String) =
+        try
+          superClass.isAssignableFrom(Class.forName(c))
+        catch {
+          case _: ClassNotFoundException =>
+            false
+        }
+
+      def loop(p: Projection): Either[Projection, Option[Projection]] =
+        p match {
+          case TypeReference(tpe) =>
+            Either.cond(!isSubclass(tpe), None, p)
+          case p @ Property(path, name, tpe) =>
+            loop(path) match {
+              case Left(path) =>
```

Comment:
can we use `Left(_) => ` here since we are not using `path`?

Reply:
done